MLflow deployment integration

stefannica opened the pull request:
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
Integration of local MLflow model server deployments and generic support for basic service tracking for ZenML

schustmi commented on src/zenml/services/service_registry.py:
```diff
@@ -0,0 +1,179 @@
+            `BaseService` subclass that was registered for the service type or
+            None, if no service class was registered for the service type.
+        """
+        # Check whether the type is registered
```
I guess this comment can be removed

schustmi commented on src/zenml/services/service_registry.py:
```diff
@@ -0,0 +1,179 @@
+        # Check whether the type is registered
+        return self.service_types.get(service_type)
+
+    def get_service_types(
```
Can we remove this method and simply use the service_types attribute instead? Or return a copy of the dict similar to `get_services`?

schustmi commented on src/zenml/services/service_registry.py:
```diff
@@ -0,0 +1,179 @@
+            self.service_types[service_type] = cls
+            logger.debug(
+                f"Registered service class {cls} for "
+                f"service type <{service_type}>"
```
Some logs specifying a service type don't include `<...>`, might be nice to remove it here as well to stay consistent (I think we use backticks instead in the remaining code base)

schustmi commented on src/zenml/services/service_registry.py:
```diff
@@ -0,0 +1,179 @@
+                f"service type <{service_type}>"
+            )
+        else:
+            raise Exception(
```
I think it would be nice to have a more concrete exception class here

stefannica replied:
Agreed. I'll use TypeError instead (seems fit, as this is triggered when declaring a service type).

schustmi commented on src/zenml/services/service_registry.py:
```diff
@@ -0,0 +1,179 @@
+        """
+        service_type = service.SERVICE_TYPE
+        if service_type not in self.service_types:
+            raise TypeError("Service type <{service_type}> is not registered.")
```
```suggestion
            raise TypeError(f"Service type `{service_type}` is not registered.")
```

schustmi commented on src/zenml/services/service.py:
```diff
@@ -0,0 +1,358 @@
+        description: description of the service
+    """
+
+    # TODO: pipeline metadata (name, run id, step etc)
```
Can we remove this or add a priority?

stefannica replied:
I'll add a priority

schustmi commented on src/zenml/services/service.py:
```diff
@@ -0,0 +1,358 @@
+    2. ensuring BaseService instance uniqueness by enforcing that no two
+    service instances have the same UUID value. Implementing this at the
+    constructor level guarantees that deserializing a service instance from
+    a JSON representation multiple times always return the same service object.
```
```suggestion
    a JSON representation multiple times always returns the same service object.
```

schustmi commented on src/zenml/services/service.py:
```diff
@@ -0,0 +1,358 @@
+        """Validate the creation of a service."""
+        if not getattr(cls, "SERVICE_TYPE", None):
+            raise RuntimeError(
+                f"Untyped services instances are not allowed. Please set the "
```
```suggestion
                f"Untyped service instances are not allowed. Please set the "
```

schustmi commented on src/zenml/services/service.py:
```diff
@@ -0,0 +1,358 @@
+    def __call__(cls, *args: Any, **kwargs: Any) -> "BaseServiceMeta":
+        """Validate the creation of a service."""
+        if not getattr(cls, "SERVICE_TYPE", None):
+            raise RuntimeError(
```
```suggestion
            raise AttributeError(
```

schustmi commented on src/zenml/services/service.py:
```diff
@@ -0,0 +1,358 @@
+        if timeout > 0:
+            self.poll_service_status(timeout)
+            if not self.is_running:
+                raise Exception(f"Failed to start service {self}.")
```
Maybe a RuntimeError instead?

schustmi reviewed (changes requested):
Incredible work on this whole PR! I'll finish my review tomorrow (just had a more or less quick look at everything apart from the LocalService and MLFlow service) but wanted to give some initial feedback already

stefannica commented on examples/mlflow_deployment/README.md:
````diff
@@ -0,0 +1,113 @@
+configuration steps, just run the following:
+
+```shell
+zenml example run mlflow_deployment
````
I see three solutions to this:
1. ignore the problem and just document in the README that the example will leave orphan daemons behind
2. remove this section from the README. Quick, but that won't solve the problem if users decide to use `zenml example run` on their own.
3. add something like a `post_run` function in `setup.sh` and change the code to handle that

I went with number 3. It was easier than I thought.

stefannica commented on examples/mlflow_deployment/pipeline.py:
```diff
@@ -0,0 +1,217 @@
+            return None
+        step = pipeline.runs[-2].get_step(name=step_name)
+        for artifact_view in step.outputs.values():
+            # filter out anything but service artifacts
```
You're right, `step.output` is enough.

stefannica commented on examples/mlflow_deployment/pipeline.py:
```diff
@@ -0,0 +1,217 @@
+
+    model_uri = mlflow.get_artifact_uri("model")
+    if not model_uri:
+        # an MLflow model was not found in the current run, so we simply reuse
```
It shouldn't happen with this particular example, because an MLflow model is always trained during the first pipeline run and then there will always be a last_service in all subsequent pipeline runs. However, raising an exception can't hurt.

stefannica commented on examples/mlflow_deployment/run.py:
```diff
@@ -0,0 +1,109 @@
+    pipe = repo.get_pipelines()[-1]
+    step = pipe.runs[-1].get_step(name=step_name)
+    for artifact_name, artifact_view in step.outputs.items():
+        # filter out anything but service artifacts
```
Yes, but in this case this is meant to be a reusable function that looks at any step with the given name and extracts the service (if any). Which reminds that I have to return None if no service is found.

stefannica commented on src/zenml/integrations/mlflow/mlflow_environment.py:
```diff
@@ -15,7 +15,15 @@
 from typing import Optional
 
-from mlflow import set_tracking_uri  # type: ignore
+from mlflow import (  # type: ignore
```
Agreed

stefannica commented on src/zenml/utils/typed_model.py:
````diff
@@ -0,0 +1,76 @@
+    It can also facilitate de-serializing objects when their type isn't known:
+
+    ```python
+    type = d['type'].split('.')
+    module = importlib.import_module('.'.join(type[:-1]))
+    cls = getattr(module, type[-1])
+    new_matrix = cls.parse_obj(d)
+    assert isinstance(new_matrix.choice, RedPill)
````
Great idea. I'll move those methods one level up.

schustmi commented on src/zenml/services/service_registry.py:
```diff
@@ -0,0 +1,179 @@
+                f"Cannot load service with unregistered service "
+                f"type: {service_type}"
+            )
+        service = service_class.from_dict(service_dict)
```
I'm a little confused how this works with pydantic, this `service_dict` still contains the service type which isn't part of the pydantic model right? Shouldn't we use `.pop(...)` in line 150?

stefannica replied:
This was quite true, with the implementation at the time you reviewed these changes. But now, I'm adding `service_type` as an actual attribute part of the pydantic model, to replace all those `to_dict/to_json/from_dict/from_json` methods.

schustmi commented on src/zenml/services/local/local_service_endpoint.py:
```diff
@@ -0,0 +1,157 @@
+            could be otherwise allocated.
+        """
+
+        # If a port value is explicitly configured, attempt to use it first
```
Similar methods to find an available port in utils/networking_utils.py, do you think we can unify them?

stefannica replied:
Yes, I'll give it a shot, although I found that the approach used in networking_utils (trying to bind the local port to see if it's taken) does not work if the port has just been released by a process.

schustmi commented on src/zenml/services/local/local_service_endpoint.py:
```diff
@@ -0,0 +1,157 @@
+                raise IOError(f"TCP port {self.config.port} is not available.")
+
+        # Attempt to reuse the port used when the services was last running
+        if self.status.port and self.port_is_available(self.status.port):
```
Essentially status.port is the one that was last used and config.port is a new configuration? I'm a little confused by the port on both objects

schustmi commented on src/zenml/services/local/local_service.py:
```diff
@@ -0,0 +1,335 @@
+    silent_daemon: bool = False
+
+    @property
+    def config_file(self) -> str:
```
I think these should be `Optional[str]` instead of returning an empty string (same for remaining properties)

schustmi commented on src/zenml/integrations/mlflow/services/mlflow_deployment.py:
```diff
@@ -0,0 +1,190 @@
+from typing import Any, Dict, Optional, Union
+
+import numpy as np
+import requests  # type: ignore
```
specific error code

schustmi commented on src/zenml/services/local/local_service.py:
```diff
@@ -0,0 +1,335 @@
+        return ServiceState.ACTIVE, ""
+
+    def _get_daemon_cmd(self) -> Tuple[List[str], Dict[str, str]]:
+        """Get the command to run to launch the service daemon.
```
```suggestion
        """Get the command to run the service daemon.
```

schustmi commented on src/zenml/services/local/local_service.py:
```diff
@@ -0,0 +1,335 @@
+
+          * the configuration describing this LocalDaemonService instance
+          is serialized as JSON and saved to a temporary file
+          * the python executable wrapper script is
```
the sentence here is a little incoherent and wrongly indented

strickvl replied:
```suggestion
          * the Python executable wrapper script is
```

schustmi commented on src/zenml/services/local/local_service.py:
```diff
@@ -0,0 +1,335 @@
+            command += ["--log-file", self.status.log_file]
+
+        command_env = os.environ.copy()
+        # command_env[_SERVER_MODEL_PATH] = local_uri
```
can probably be removed

schustmi commented on src/zenml/integrations/mlflow/services/mlflow_deployment.py:
```diff
@@ -0,0 +1,190 @@
+
+    model_uri: str
+    model_name: str
+    workers: Optional[int] = 1
```
Can these be non-optional if we have default values for them?

stefannica commented on src/zenml/utils/typed_model.py:
```diff
@@ -0,0 +1,76 @@
+    def __new__(
+        mcs, name: str, bases: Tuple[Type[Any], ...], dct: Dict[str, Any]
+    ) -> "BaseTypedModelMeta":
+        """Creates a pydantic BaseModel class that includes a hidden attribute that
```
I wasn't aware of that. Definitely worth using.

strickvl commented on src/zenml/integrations/mlflow/mlflow_environment.py:
```diff
@@ -98,3 +107,107 @@ def deactivate(self) -> None:
+
+class MLFlowStepEnvironment(BaseEnvironmentComponent):
+    """Provides information about an MLflow step environment.
+    To access it inside your step function:
```
As an aside / comment, I really like these code-inside-docstrings examples. I think we should do it throughout our codebase where possible.

strickvl commented on src/zenml/integrations/mlflow/mlflow_environment.py:
```diff
@@ -98,3 +107,107 @@ def deactivate(self) -> None:
+    def _create_or_reuse_mlflow_run(self) -> None:
+        """Create or reuse an MLflow run for the current step.
+
+        IMPORTANT: this function is not race condition proof. If two or more
```
```suggestion
        IMPORTANT: this function might cause a race condition. If two or more
```

Possibly a bit clearer? Feel free to ignore this one.

strickvl commented on src/zenml/integrations/mlflow/services/mlflow_deployment.py:
```diff
@@ -0,0 +1,190 @@
+        workers: number of workers to use for the prediction service
+        mlserver: set to True to use the MLflow MLServer backend (see
+            https://github.com/SeldonIO/MLServer). If False, the
+            MLflow builtin scoring server will be used.
```
```suggestion
            MLflow built-in scoring server will be used.
```

strickvl commented on src/zenml/services/local/local_service.py:
```diff
@@ -0,0 +1,335 @@
+            daemon output will be redirected to a logfile.
+    """
+
+    silent_daemon: bool = False
```
Dumb question: I wasn't fully sure how the `silent_daemon` property on the `LocalDaemonServiceConfig` interacts / works in terms of hierarchy with the same property on `LocalDaemonServiceStatus`. Might this end up being confusing for the user?

stefannica replied:
This is an awkward consequence of the fact that I need to access the configuration object in the status object, but I can't keep a link to it. There is a similar case with the port field in the local endpoint status. Fixing this will require some refactoring changes, so for now I would just add a TODO for it.

strickvl commented on src/zenml/services/local/local_service.py:
```diff
@@ -0,0 +1,335 @@
+          is serialized as JSON and saved to a temporary file
+          * the python executable wrapper script is
+          the configuration is to launch a
+        python wrapper that instantiates the same LocalDaemonService object
```
```suggestion
        Python wrapper that instantiates the same LocalDaemonService object
```

strickvl commented on src/zenml/services/local/local_service.py:
```diff
@@ -0,0 +1,335 @@
+
+        Returns:
+            Command needed to launch the daemon process and the environment
+            variables to set for it, in the formats accepted by
```
```suggestion
            variables to set, in the formats accepted by
```

strickvl commented on src/zenml/services/local/local_service_endpoint.py:
```diff
@@ -0,0 +1,157 @@
+            port: TCP port number
+
+        Returns:
+            True if the port is available, False otherwise
```
```suggestion
            True if the port is available, otherwise False
```

strickvl commented on src/zenml/services/local/local_service_endpoint.py:
```diff
@@ -0,0 +1,157 @@
+            An available TCP port number
+
+        Raises:
+            IOError: if the preferred TCP port is busy and `allocate_port` is
+            disabled in the endpoint configuration, or if no free TCP port
+            could be otherwise allocated.
```
I think indentations here are off?

strickvl commented on src/zenml/services/service.py:
```diff
@@ -0,0 +1,358 @@
+
+    This class implements generic functionality concerning the life-cycle
+    management and tracking of an external service (e.g. process, container,
+    kubernetes deployment etc.).
```
```suggestion
    Kubernetes deployment etc.).
```

strickvl commented on src/zenml/services/service.py:
```diff
@@ -0,0 +1,358 @@
+        Returns:
+            The operational state of the external service and a message
+            providing additional information about that state (e.g. a
+            description of the error, if one is encountered while checking the
```
```suggestion
            description of the error if one is encountered while checking the
```

strickvl commented on src/zenml/services/service.py:
```diff
@@ -0,0 +1,358 @@
+
+    def poll_service_status(self, timeout: int = 0) -> None:
+        """Poll the external service status until the service operational
+        state matches the administrative state, or the timeout is reached.
```
```suggestion
        state matches the administrative state or the timeout is reached.
```

strickvl commented on src/zenml/services/service.py:
```diff
@@ -0,0 +1,358 @@
+
+        Returns:
+            True if the service is running and active (i.e. the endpoints are
+            responsive, if any are configured), False otherwise.
```
```suggestion
            responsive, if any are configured), otherwise False.
```

strickvl commented on src/zenml/services/service.py:
```diff
@@ -0,0 +1,358 @@
+        if timeout > 0:
+            self.poll_service_status(timeout)
+            if not self.is_stopped:
+                raise Exception(f"Failed to stop service {self}.")
```
Here also maybe a `RuntimeError` like @schustmi suggested above?

strickvl commented on src/zenml/services/service_endpoint.py:
```diff
@@ -0,0 +1,172 @@
+            # port and protocol are known
+            return None
+
+        return f"{self.protocol.value}://{self.hostname}:{self.port}"
```
Not sure of the tradeoffs or what is usually done here, but would it maybe be a good idea to end with a `/` for this returned value? Thinking about future use cases where people are building on top of these endpoints with query strings?

strickvl commented on src/zenml/services/service_endpoint.py:
```diff
@@ -0,0 +1,172 @@
+
+    def update_status(self) -> None:
+        """Check the the current operational state of the external service
+        endpoint and update the local operational status information to reflect
+        it.
```
```suggestion
        endpoint and update the local operational status information accordingly.
```

strickvl commented on src/zenml/services/service_endpoint.py:
```diff
@@ -0,0 +1,172 @@
+        endpoint status and will return the result.
+
+        Returns:
+            True if the service endpoint is active, False otherwise.
```
```suggestion
            True if the service endpoint is active, otherwise False.
```

strickvl commented on src/zenml/services/service_endpoint.py:
```diff
@@ -0,0 +1,172 @@
+        endpoint status and will return the result.
+
+        Returns:
+            True if the service endpoint is inactive, False otherwise.
```
```suggestion
            True if the service endpoint is inactive, otherwise False.
```

strickvl commented on src/zenml/services/service_monitor.py:
```diff
@@ -0,0 +1,228 @@
+    """Base class used for service endpoint health monitors.
+
+    Attributes:
+        config: endpoint health monitor configuration
```
```suggestion
        config: health monitor configuration for endpoint
```

strickvl commented on src/zenml/services/service_monitor.py:
```diff
@@ -0,0 +1,228 @@
+    """HTTP service endpoint health monitor.
+
+    Attributes:
+        config: HTTP endpoint health monitor configuration
```
```suggestion
        config: health monitor configuration for HTTP endpoint
```

strickvl commented on src/zenml/services/service_monitor.py:
```diff
@@ -0,0 +1,228 @@
+    """TCP service endpoint health monitor.
+
+    Attributes:
+        config: TCP endpoint health monitor configuration
```
```suggestion
        config: health monitor configuration for TCP endpoint
```

strickvl commented on src/zenml/services/service_registry.py:
```diff
@@ -0,0 +1,179 @@
+        """Registers a new service type.
+
+        Args:
+            cls: A BaseService subclass.
```
```suggestion
            cls: a BaseService subclass.
```

strickvl commented on src/zenml/services/service_registry.py:
```diff
@@ -0,0 +1,179 @@
+        """Registers a new service instance
+
+        Args:
+            service: A BaseService instance.
```
```suggestion
            service: a BaseService instance
```

strickvl commented on src/zenml/services/service_status.py:
```diff
@@ -0,0 +1,75 @@
+
+    ACTIVE = "active"
+    PENDING_STARTUP = "pending_startup"
+    INACTIVE = "inactive"
```
Dumb question: what defines an inactive service? How do we define whether something is inactive?

stefannica replied:
If it helps, I do think these states need some refactoring too, to make them more user-friendly, so if you have any suggestions about that, I welcome them.

strickvl commented on src/zenml/services/service_status.py:
```diff
@@ -0,0 +1,75 @@
+
+class ServiceStatus(BaseModel):
+    """Information describing the operational status of an external process
+    or service tracked by ZenML (e.g. process, container, kubernetes
```
```suggestion
    or service tracked by ZenML (e.g. process, container, Kubernetes
```

strickvl commented on src/zenml/utils/daemon.py:
```diff
@@ -33,19 +33,150 @@
+def daemonize(
+    _func: F,
+) -> F:
+    """Type annotations for deamonizer decorator in case of no arguments."""
```
```suggestion
    """Type annotations for daemonizer decorator in case of no arguments."""
```

strickvl commented on src/zenml/utils/daemon.py:
```diff
@@ -33,19 +33,150 @@
+    pid_file: Optional[str] = None,
+    working_directory: str = "/",
+) -> Callable[[F], F]:
+    """Type annotations for deamonizer decorator in case of arguments."""
```
```suggestion
    """Type annotations for daemonizer decorator in case of arguments."""
```

strickvl commented on src/zenml/utils/daemon.py:
```diff
@@ -33,19 +33,150 @@
+            its stdout/stderr redirected to /dev/null).
+        pid_file: an optional file where the PID of the daemon process will
+            be stored.
+        working_directory: Working directory for the daemon process,
```
```suggestion
        working_directory: working directory for the daemon process,
```

strickvl commented on src/zenml/utils/typed_model.py:
```diff
@@ -0,0 +1,76 @@
+
+
+class BaseTypedModelMeta(ModelMetaclass):
+    """Metaclass responsible for adding type information to pydantic Models."""
```
```suggestion
    """Metaclass responsible for adding type information to Pydantic models."""
```

strickvl commented on src/zenml/utils/typed_model.py:
```diff
@@ -0,0 +1,76 @@
+
+
+class BaseTypedModel(BaseModel, metaclass=BaseTypedModelMeta):
+    """Type pydantic Model base class.
```
```suggestion
    """Typed Pydantic model base class.
```

strickvl commented on tests/integration/test_examples.py:
```diff
@@ -243,6 +247,82 @@ def test_run_mlflow(examples_dir: Path):
+
+    from zenml.integrations.mlflow.mlflow_environment import MLFlowEnvironment
+
+    # Create and activate the global MLflow environment
```
```suggestion
    # create and activate the global MLflow environment
```

stefannica commented on src/zenml/utils/daemon.py:
```diff
@@ -33,19 +33,150 @@
-        pid_file: str,
+        daemon_function: F,
+        *args: Any,
+        pid_file: Optional[str] = None,
```
This is just me being overzealous with flexibility. But you're right, daemons should always have an associated PID file. I'll change this back.

stefannica commented on src/zenml/services/service_status.py:
```diff
@@ -0,0 +1,75 @@
+    ERROR = "error"
+
+
+class ServiceStatus(BaseModel):
```
No, just a mistake on my part. Good catch !

stefannica commented on src/zenml/services/service_monitor.py:
```diff
@@ -0,0 +1,228 @@
+    )
+
+    @classmethod
+    def port_is_open(cls, hostname: str, port: int) -> bool:
```
Great idea ! I found some other similar utility functions there too (TCP port availability, allocation). I'll point the code to them rather than duplicating everything.

stefannica commented on src/zenml/integrations/kubeflow/orchestrators/local_deployment_utils.py:
```diff
@@ -297,5 +297,5 @@ def stop_kfp_ui_daemon(pid_file_path: str) -> None:
             from zenml.utils import daemon
 
-            daemon.stop_daemon(pid_file_path, kill_children=True)
+            daemon.stop_daemon(pid_file_path)
```
Same here: the daemon itself is now capable of killing its own children when terminated.

stefannica commented on src/zenml/services/service_endpoint.py:
```diff
@@ -0,0 +1,172 @@
+        port: the current TCP port where the service endpoint is accessible
+    """
+
+    protocol: ServiceEndpointProtocol = ServiceEndpointProtocol.TCP
```
For the local endpoint case, yes, these certainly feel like they belong in the configuration rather than status, because on the localhost these resources (hostname, port, protocol) are known or can be configured in advance. However, for the general case where the service runs somewhere remotely (e.g. in a managed Kubernetes cluster), these resources are dynamically allocated when the service is provisioned.

stefannica commented on src/zenml/services/service_endpoint.py:
```diff
@@ -0,0 +1,172 @@
+    """
+
+    admin_state: ServiceState = ServiceState.INACTIVE
+    config: ServiceEndpointConfig = Field(default_factory=ServiceEndpointConfig)
```
I thought I needed a default factory to not be forced to supply a config and a status when creating a ServiceEndpoint, but I checked and these aren't really needed. What I wanted here is to have default configuration and (especially) status, to make it easier for users to instantiate these classes.

I removed the default factories.

stefannica commented on src/zenml/services/service_endpoint.py:
```diff
@@ -0,0 +1,172 @@
+        super().__init__(*args, **kwargs)
+        self.config.name = self.config.name or self.__class__.__name__
+
+    def check_status(self) -> Tuple[ServiceState, str]:
```
I wanted to have a clean separation of concerns here:
1. `check_status` is implemented by a concrete service that checks the external resource (process, container, Kubernetes deployment, VM, managed public cloud resource/service, etc.) and returns the minimum amount of information that is needed by 2. to implement that operational state machine: the current operational state (`ServiceState`) and an optional user-friendly message describing that state.
2. `poll_service_status` is where the base service implementation provides _an operational state machine_ that looks at the `admin_state` and the operational state returned by `check_status` and determines what to do next with the service (e.g. provision the service if it's stopped or in an error state and the admin state is `active` or give up after some number of retries). `ServiceStatus` is where the operational machine state stores the full state of the result that also takes into consideration the admin state.

An example where this separation is useful: `check_status` returns an INACTIVE state, but the _operational state machine_ interprets this differently depending on what the service is doing - either the service has successfully shut down, or the service is still being provisioned.

Either way, I need/plan to go back to this _operational state machine_ and make it more robust in the future. Right now it's too simple, but sufficient for an MVP.

stefannica commented on src/zenml/services/service.py:
```diff
@@ -0,0 +1,358 @@
+        if self.endpoint:
+            self.endpoint.update_status()
+
+    def to_dict(self) -> Dict[str, Any]:
```
I couldn't really use pydantic's `dict()` to also make `json()` work automatically, because `json()` doesn't call `dict()` under the hood, but I did manage to remove all 4 methods from the BaseService class and replace them by adding a `service_type` immutable pydantic attribute to the class automatically with the meta-class.

stefannica commented on src/zenml/services/local/local_service.py:
```diff
@@ -0,0 +1,335 @@
+            )
+            return
+
+        logger.info("Stopping daemon for service '%s' ...", self)
```
Great idea. I added `console.status` rich support, but not here, in the BaseService class start/stop_service, where the waiting happens.

stefannica commented on src/zenml/services/local/local_service.py:
```diff
@@ -0,0 +1,335 @@
+            )
+            return
+
+        logger.info("Starting daemon for service '%s'...", self)
```
Same as above. The result looks great, too :)

stefannica commented:
@strickvl @schustmi I addressed all your remaining review comments with the last 3 commits. I also pimped up the example to be more relevant from an end-user perspective (e.g. there are two pipelines now, one for continuous deployment, the other one for inference).

strickvl commented on examples/mlflow_deployment/README.md:
```diff
@@ -66,16 +85,24 @@ python run.py
+will use the currently running MLflow deployment server to perform an
+online prediction.
+
+Similartly, if a new model is trained in the deployment pipeline but the
```
Should be 'similarly'

strickvl reviewed (changes requested):
Just one typo I saw...

schustmi commented on examples/mlflow_deployment/setup.sh:
```diff
@@ -0,0 +1,18 @@
+  zenml integration install mlflow -f
+}
+
+# post_run () {
```
Any reason this isn't uncommented?

schustmi reviewed (changes requested):
Really like the new example! Just found one small part that needs some clarification

schustmi reviewed (approved):
🥳

stefannica commented:
Squashed and rebased on top of develop.
